### ISSUE TITLE
Graph transformations in the fit-transform paradigm

### ISSUE THREAD

**Issue by davidfstein** [pyg-team/pytorch_geometric#6636]
### 🚀 The feature, motivation and pitch

In some scenarios we need to preprocess data such as node features belonging to our graphs. In a supervised setting, this requires that we "fit" a transformation (e.g. normalization, imputation) to the data from our training set and apply the fit transformation to the training and testing data to avoid information leakage. Currently, you implement transformations such as "NormalizeFeatures". However, from the documentation I can not see a way to apply these transformations as described above. Would it be possible to implement a scikit-learn like interface with "fit" and "transform" functions for these transformations? If not could you recommend a way to achieve properly isolated preprocessing? Currently, I am extracting the features from the Dataset object, transforming them and then setting them. This does not seem like the optimal solution and sometimes behaves strangely.

**Comment by rusty1s** [comment]
This is an interesting topic. Currently, I don't think we have many transforms that are data-dependent, e.g., where a fit/transform paradigm would be necessary.

I see that `torchvision` just requires stats to be pre-computed, see [here](https://github.com/pytorch/vision/blob/main/torchvision/transforms/functional.py#L340). Is that something that would be already sufficient?

**Comment by davidfstein** [comment]
I think in the case of normalization that would probably work, but for other types of transformations such as encoding categorical data I'm not sure it would be sufficient. Actually, the scikit style interface isn't really necessary and comes along with a lot of design choices that maybe don't fit with pyg. All thats really needed is a way to set the attributes of a dataset. Then the user can do whatever they want. 

Right now this is what I've had to do,
`x = torch.vstack([data.x for data in train_dataset])
scaler = StandardScaler()
scaler.fit(x)
train_dataset.data.x = torch.tensor(scaler.transform(train_dataset.data.x.numpy())).float()
test_dataset.data.x = torch.tensor(scaler.transform(test_dataset.data.x.numpy())).float()
scaled_data = []
for data in tqdm(train_dataset):
    data.x = torch.tensor(scaler.transform(data.x)).float()
    scaled_data.append(data)
train_dataset._data_list = scaled_data`

For some reason, I have to set the train dataset's x attribute and _data_list attribute separately in order for them to be equal. For the test dataset however setting the x attribute is sufficient. This doesn't feel like a good way to do this and probably will break easily. If there is already a better way to achieve this then maybe my problem is already solved

**Comment by rusty1s** [comment]
I think your approach looks reasonable. We could think of integrating such an interface to apply `scikit`-style like transforms.

Regarding your issue: Yeah, this is a bit tricky. `[data.x ...]` will cache the results in `_data_list` and so you need to modify `_data_list` as well for the training set. At best, just do `x = train_dataset.data.x` so that no caching will take place.

**Comment by davidfstein** [comment]
Ah ok thats good to know. It might be nice to implement a setter. Because some may try this without setting _data_list and end up with surprising results. In the setter, you could check that the x and _data_list objects are in sync in the scenario when results are cached

**Comment by rusty1s** [comment]
Actually, in master we now print a warning in this case. It might be still good to clear the cache whenever someone updates `data.x`.

**Comment by davidfstein** [comment]
Oh nice, that should make things safer. I think in that case, my issue should be solved. I can close unless you want to keep it open for a pr to clear the cache

**Comment by rusty1s** [comment]
Yes, let's keep it open till then.
